[DOC] developer documentation for the testing framework (#1922)
alan-turing-institute/sktime: docs/source/developer_guide/testing_framework.rst

instead of custom logic, see `pytest documentation <pytestuse>`_.

Estimator tests use ``sktime``'s framework plug-in to ``pytest_generate_tests``,
which parameterizes estimator fixtures and data input scenarios.

An illustrative example
~~~~~~~~~~~~~~~~~~~~~~~

Starting with an example:

.. code-block::

   def test_fit_returns_self(estimator_instance, scenario):
      """Check that fit returns self."""
      fit_return = scenario.run(estimator_instance, method_sequence=["fit"])
      assert (
         fit_return is estimator_instance
      ), f"Estimator: {estimator_instance} does not return self when calling fit"

Here, ``estimator_instance`` and ``scenario`` fixture are looped over, assuming
values from a list of fixtures:
* ``estimator_instance`` over estimator instances, obtained from all ``sktime`` estimators via ``create_test_instances_and_names``
* ``scenario`` objects, which encodes data inputs and method call sequences to ``estimator_instance`` (explained in further detail below).

The ``scenario.run`` command is equivalent to calling ``estimator_instance.fit(**scenario_kwargs)``,
where the ``scenario_kwargs`` are generated by the ``scenario``.

> It should be noted that the test is not decorated decorated with fixture parameterization,

Author: both spellings for parameterization are fine, but why not.

the fixtures are instead generated by ``pytest_generate_tests``.

The reason for this is that the applicable scenarios (fixture values of ``scenario``) depend on the ``estimator_instance`` fixture,
since inputs to ``fit`` of a classifier will differ to an input to ``fit`` of a forecaster.

Parameterized fixtures
~~~~~~~~~~~~~~~~~~~~~~

> Currently, the ``sktime`` testing framework parameterizes the following fixtures:

Author: it's a standing term in testing? See also https://docs.pytest.org/en/6.2.x/parametrize.html
Or, could you concretely explain which term or formulation is hard to understand?

> Currently, the ``sktime`` testing framework parameterizes the following fixtures:

Author: A paragraph has been added which references the `pytest` documentation on terms and concepts.

> 

Author: yes, ok

> 

Author: clarified this, better?

* ``estimator``: all estimator classes, inheriting from the base class of the given module.
In ``test_all_estimators``, this loops over all estimators.
* ``estimator_instance``: all estimator test instances, obtained from all ``sktime`` estimators via ``create_test_instances_and_names``
* ``scenario``: test scenarios, applicable to ``estimator`` or ``estimator_instance``.
   The scenarios are specified in ``utils/_testing/scenarios_[estimator_scitype]``.

Scenarios
~~~~~~~~~

> The ``scenario`` fixtures encode arguments for method calls, and a sequence for method calls.

Reviewer: not sure if there's any actual encoding happening

```suggestion
The ``scenario`` fixtures contain arguments for method calls, and a sequence for method calls.
```

Author: well, sort of yes, but in any case "contain" is clearer.

An example scenario specification, from ``utils/_testing/scenarios_forecasting``:

.. code-block::

   class ForecasterFitPredictUnivariateNoXLateFh(ForecasterTestScenario):
      """Fit/predict only, univariate y, no X, no fh in predict."""

      _tags = {"univariate_y": True, "fh_passed_in_fit": False}

      args = {
         "fit": {"y": _make_series(n_timepoints=20, random_state=RAND_SEED)},
         "predict": {"fh": 1},
      }
      default_method_sequence = ["fit", "predict"]

The scenario ``ForecasterFitPredictUnivariateNoXLateFh`` encodes instructions
applied to an ``estimator_instance``, via instances ``scenario``.
A call ``result = scenario.run(estimator_instance)`` will:

1. first, call ``estimator_instance.fit(y=_make_series(n_timepoints=20, random_state=RAND_SEED))``
2. then, call ``estimator_instance.predict(fh=1)`` and return the  output too ``result``.

The abstraction of "scenario" allows to easily specify multiple argument combinations across multiple methods.

> The method ``run`` also has arguments that allow to override the method sequence, e.g.,

Reviewer: Maybe mention the arguments that allow for that behavior explicitly

Author: added

run them in a different order, or only a subset thereof.

Scenarios also provide a method ``scenario.is_applicable(estimator)``, which returns a boolean, whether
``scenario`` is applicable to ``estimator``. For instance, scenarios with univariate data are not applicable
to multivariate forecasters, and will cause exceptions in a ``fit`` method call.
Non-applicable scenarios can be filtered out in positive tests, and filtered in in negative tests.
As a default, the ``sktime`` implemented ``pytest_generate_tests`` only pass applicable scenarios.

> Further, scenarios inherit from ``BaseObject``, which allows to ue the ``sktime`` tag system with scenarios.

Reviewer: ```suggestion
Further, scenarios inherit from ``BaseObject``, which allows to use the ``sktime`` tag system with scenarios.
```

> For further details on scenarios, inspect the docstring of ``BaseScenario``.

Reviewer: maybe link to the module explicitly so that people can find it quickly

Author: I was hesitant since the location may change in case we refactor the random heap of things that the `utils` module is...

Author: link added

 For instance, ``forecasting/tests/test_all_forecasters.py``, or ``distances/tests/test_all_dist_kernels.py``.
* Learning task specific tests should not duplicate package level, generic estimator tests in ``test_all_estimators.py``

Test code architecture
----------------------

.. _pytestuse: https://docs.pytest.org/en/6.2.x/example/index.html

``sktime`` test files should use best ``pytest`` practice such as fixtures or test parameterization where possible,
instead of custom logic, see `pytest documentation <pytestuse>`_.

Estimator tests use ``sktime``'s framework plug-in to ``pytest_generate_tests``,
which parameterizes estimator fixtures and data input scenarios.

An illustrative example
~~~~~~~~~~~~~~~~~~~~~~~

Starting with an example:

.. code-block::

   def test_fit_returns_self(estimator_instance, scenario):
      """Check that fit returns self."""
      fit_return = scenario.run(estimator_instance, method_sequence=["fit"])
      assert (
         fit_return is estimator_instance
      ), f"Estimator: {estimator_instance} does not return self when calling fit"

> Here, ``estimator_instance`` and ``scenario`` fixture are looped over, assuming

Reviewer: Where exactly is the loop?

Author: the loop is carried out by `pytest`, via fixture paramterization (see there)

Author: added more explanation here

values from a list of fixtures:
* ``estimator_instance`` over estimator instances, obtained from all ``sktime`` estimators via ``create_test_instances_and_names``
* ``scenario`` objects, which encodes data inputs and method call sequences to ``estimator_instance`` (explained in further detail below).

The ``scenario.run`` command is equivalent to calling ``estimator_instance.fit(**scenario_kwargs)``,
where the ``scenario_kwargs`` are generated by the ``scenario``.

> It should be noted that the test is not decorated with fixture parametrization,

Reviewer: Does "the test" refer to the test above? what about "the fixtures"?

Author: to the example code block, I'll clarify

the fixtures are instead generated by ``pytest_generate_tests``.

The reason for this is that the applicable scenarios (fixture values of ``scenario``) depend on the ``estimator_instance`` fixture,
since inputs to ``fit`` of a classifier will differ to an input to ``fit`` of a forecaster.

Parameterized fixtures
~~~~~~~~~~~~~~~~~~~~~~

.. _pytestfixtparam: https://docs.pytest.org/en/6.2.x/parametrize.html

``sktime`` uses ``pytest`` fixture parameterization to execute tests in a loop over fixtures,
for instance running all interface compatibility tests for all estimators.
See the `pytest documentation on fixture parameterization <pytestfixtparam>`_ for an explanation of fixture parameterization.

Currently, the ``sktime`` testing framework parameterizes the following fixtures in module level tests:

> * ``estimator``: all estimator classes, inheriting from the base class of the given module.

Reviewer: This is quite hard to read, would it be possible to rephrase? there is "testing framework parameterizes the following fixtures" and there is only 1 element after colon... Is it an example or a comprehensive list of length 1?

Author: no? As with fixture parameterization, there is the fixture variable (`estimator`), and the value range over which the parameterization takes place. The value range is the set/collection of all estimator classes, of which there are many (examples: ARIMA forecaster, HIVE-COTE classifier, dtw-aligner)

Author: any suggestions how to improve this?

Author: gave it a try

> In ``test_all_estimators``, this loops over all estimators.

Reviewer: Does "this loops" refer to testing  framework, scenarios, test collection or something else?

Author: perhaps better: "In the package level tests `test_all_estimators`, that base class is `BaseEstimator`"

Reviewer: Yup that is better

Author: adopted

> * ``estimator_instance``: all estimator test instances, obtained from all ``sktime`` estimators via ``create_test_instances_and_names``

Reviewer: should I read "estimator test instances" as "estimator instances" or "test instances"?

Author: well, the test instances are estimator instances, and they are returned by `create_test_instances_and_names`. Does that help, is it unclear?

Reviewer: ```suggestion
* ``estimator_instance``: all estimator instances from  ``create_test_instances_and_names``
```

Author: that is less clear (this is a method, not a function!), so I think this is worse than the original.

Reviewer: I don't think I understand what you mean in that sentence and in your comment, up to you to leave it as is

* ``scenario``: test scenarios, applicable to ``estimator`` or ``estimator_instance``.
   The scenarios are specified in ``utils/_testing/scenarios_[estimator_scitype]``.

Further parameterization may happen for individual tests, the scope is usually explained in the test docstrings.

Scenarios
~~~~~~~~~

> The ``scenario`` fixtures contain arguments for method calls, and a sequence for method calls.

Reviewer: ```suggestion
The ``scenario`` fixtures contain  a sequence for method and arguments for each method.
```

Author: not entirely correct, so I'd rather stick with the original

Reviewer: this reads backwards to me.

An example scenario specification, from ``utils/_testing/scenarios_forecasting``:

.. code-block::

   class ForecasterFitPredictUnivariateNoXLateFh(ForecasterTestScenario):
      """Fit/predict only, univariate y, no X, no fh in predict."""

      _tags = {"univariate_y": True, "fh_passed_in_fit": False}

      args = {
         "fit": {"y": _make_series(n_timepoints=20, random_state=RAND_SEED)},
         "predict": {"fh": 1},
      }
      default_method_sequence = ["fit", "predict"]

The scenario ``ForecasterFitPredictUnivariateNoXLateFh`` encodes instructions
applied to an ``estimator_instance``, via instances ``scenario``.
A call ``result = scenario.run(estimator_instance)`` will:

1. first, call ``estimator_instance.fit(y=_make_series(n_timepoints=20, random_state=RAND_SEED))``
2. then, call ``estimator_instance.predict(fh=1)`` and return the  output too ``result``.

> The abstraction of "scenario" allows to easily specify multiple argument combinations across multiple methods.

Reviewer: Good practice is not to assume how easy/difficult things will be for the end user
```suggestion
The abstraction of "scenario" allows to specify multiple argument combinations across multiple methods.
```

The method ``run`` also has arguments (``method_sequence`` and ``arg_sequence``)
that allow to override the method sequence, e.g.,
run them in a different order, or only a subset thereof.

Scenarios also provide a method ``scenario.is_applicable(estimator)``, which returns a boolean, whether
``scenario`` is applicable to ``estimator``. For instance, scenarios with univariate data are not applicable
to multivariate forecasters, and will cause exceptions in a ``fit`` method call.
Non-applicable scenarios can be filtered out in positive tests, and filtered in in negative tests.
As a default, the ``sktime`` implemented ``pytest_generate_tests`` only pass applicable scenarios.

> Further, scenarios inherit from ``BaseObject``, which allows to use the ``sktime`` tag system with scenarios.

Reviewer: On a general note `BaseObject` is a terrible name, consider renaming to `SktimeObject` or `SktimeBaseClass`

Author: out of scope for this PR, but sure.

For further details on scenarios, inspect the docstring of ``BaseScenario``.

Extending the testing module
----------------------------

This section explains how to extend the testing module.
Depending on the primary change that is tested, the changes to the testing module will
be shallow or deep. In decreasing order of commonality:

> * When adding new estimators or utiliy functionality, it usually suffices to write only
> low level tests that check correctness of the estimator.

Reviewer: ```suggestion
* When adding new estimators or utility functionality, write low level tests that check correctness of the estimator.
```

These typically use only the simplest idioms in ``pytest`` (e.g., fixture parameterization).
Adding a new estimator will typically *not* require changes to module level tests,

> as new estimators are automatically discovered and looped over by the existing tests.

Reviewer: Great, not need for instructions what is not necessary :)

Author: hm, what would be more informative? Saying positively what needs change, primarily?

Reviewer: I would drop the those lines to simplify

Author: I changed this to a positive statement

* adding the estimator or test/estimator combination to the ``EXCLUDED_TESTS`` or ``EXCLUDE_ESTIMATORS`` in the appropriate ``_config`` file.
* adding a check condition in the ``is_excluded`` method used in ``pytest_generate_fixtures``, possibly only if the testing module supports this

Escaping tests directly in the tests, e.g., via ``if isinstance(estimator_instance, MyClass)`` should be avoided where possible.

Adding package or module level tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Module level tests use ``pytest_generate_tests`` to define fixtures.

The available fixtures vary per module, and are listed in the docstring of ``pytest_generate_tests``.

A new test should use these fixtures, if possible, but also can add new fixtures via ``pytest`` basic fixture functionality.

If new fixture variables are to be used throughout the module, or depend on existing fixtures,
instructions in the next section should be followed.

Where possible, scenarios should be used to simulate generic method calls (see above),
instead of creating and passing arguments directly. Scenarios will ensure consistent coverage of input argument cases.

Adding fixture variables
~~~~~~~~~~~~~~~~~~~~~~~~

> One-off fixture variables can be added using ``pytest`` basic functionality.

Reviewer: Not much info in this sentence, consider removing

Author: Why? This is "option 1", the rest is "option 2". Removing "option 1" removes crucial information imo?

Reviewer: I don't find any useful information in this sentence so my suggestion is to remove, but only a suggestion. Feel free to resolve if you think otherwise

Author: made it clearer that these are two cases. Better?

Fixtures used throughout module or package level tests should be added to the
fixture generation process called by ``pytest_generate_tests``.

This requires:
* adding a function ``_generate_[variablename](test_name, **kwargs)``, as described below
* assigning the function to ``generator_dict["variablename"]``
* adding the new variable in the ``fixture_sequence`` list in ``pytest_generate_tests``

> The function ``_generate_[variablename](test_name, **kwargs)`` should return two objects:

Reviewer: for consistency and readability
```suggestion
The function ``_generate_[variable_name](test_name, **kwargs)`` should return two objects:
```

> * a list of fixture to loop over, to substitute for ``variablename`` when appearing in a test signature

Reviewer: ```suggestion
* a list of fixture to loop over, to substitute for ``variable_name`` when appearing in a test signature
```

* a list of names of equal length, i-th element used as a name for the i-th fixture in test logs

The function has access to:
* ``test_name``, the name of the test the variable is called in.
This can be used to customize the list of fixtures for specific tests,
although this is meant for generic behaviour mainly.
One-off escapes and similar should be avoided here, and instead dealt with ``xfail`` and similar.
* the value of the fixture variables that appear earlier in ``fixture_sequence``, in ``kwargs``.
For instance, the value of ``estimator_instance``, if this is a variable used in the test.

> This can be used to make the list of fixtures for ``variablename`` dependent on the value of other fixtures variables

Reviewer: ```suggestion
This can be used to make the list of fixtures for ``variable_name`` dependent on the value of other fixtures variables
```
